Unstall merge topology snapshot

bhalevy opened the pull request:
This series adds facilities to gently convert canonical mutations back to mutations
and to gently make canonical mutations or freeze mutations in a seastar thread.

Those are used in storage_service::merge_topology_snapshot to prevent reactor stalls
due to large mutation, as seed in the test_add_many_nodes_under_load dtest.

Also, migration_manager migration_request was converted to use a seastar thread
to use the above facilities to prevent reactor stalls with large schema mutations,
e,g, with a large number of tables, and/or when reading tablets mutations with
a large number of tablets in a table.

perf-simple-query --write results:
Before:
```
median 79151.53 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53289 insns/op,        0 errors)
```
After:
```
median 79716.73 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53314 insns/op,        0 errors)
```

bhalevy commented:
In v2 ([664d766](https://github.com/scylladb/scylladb/commit/664d766f9f0f7bea4678b0b30f57fa4c17bed1b3))
- rebased
- made sure there's forward progress in `write_serialized_executor` if `needs_preempt` (since it returns true unconditionally in debug mode)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8148/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8148/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/canonical_mutation_test
:small_blue_diamond: boost/frozen_mutation_test
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: boost/mutation_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8148/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9109/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1087/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8148/)
#### Build Details:
- Duration: 3 hr 19 min
- Builder: spider8.cloudius-systems.com

denesb commented on replica/table.cc:
```diff
@@ -2910,23 +2909,17 @@ future<> table::apply(const mutation& m, db::rp_handle&& h, db::timeout_clock::t
     if (_virtual_writer) [[unlikely]] {
-        return (*_virtual_writer)(m);
+        co_await (*_virtual_writer)(m);
+        co_return;
```
Using coroutines here will introduce an unconditional allocation on the fast-path. I think @avikivity doesn't likes those.

avikivity replied:
absolutely, this is a very hot path, please measure with perf-simple-query --write.

denesb commented on mutation/mutation_partition.cc:
```diff
@@ -325,10 +326,12 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
                 res = apply_resume(stage, i->position());
                 return stop_iteration::no;
             }
+            */
```
Why leave this commented code?

denesb replied:
I also don't understand why adding the new apply variants requires this change.

bhalevy replied:
sorry, this was left over unintentionally, while debugging.

denesb commented on mutation/mutation_partition.cc:
```diff
@@ -462,6 +449,25 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
+        // TODO: Drop redundant range tombstones
+        p._tombstone = {};
+    }
+
```
Why this code movement?

bhalevy replied:
sorry, this was leftover from debugging as well.
although it might merit an independent change (that should be considered on its own).
the reason is to make compaction more effective.
the current code applies the tombstone asymmetrically, but we may want to compact the result of the applied partition.

denesb commented on mutation/mutation_partition.cc:
```diff
@@ -478,6 +484,22 @@ mutation_partition::apply(const schema& s, mutation_partition_view p,
+future<>
+mutation_partition::apply_gently(const schema& s, mutation_partition_view p,
+        const schema& p_schema, mutation_application_stats& app_stats) {
+    // FIXME: Optimize
```
Why re-add this FIXME, I saw you removed it in earlier commits.

denesb commented on mutation/mutation_partition_serializer.cc:
```diff
@@ -182,6 +184,95 @@ void mutation_partition_serializer::write_serialized(Writer&& writer, const sche
+            _range_tombstones->writer.add().write_start(rt.start).write_tomb(rt.tomb).write_start_kind(rt.start_kind)
+                .write_end(rt.end).write_end_kind(rt.end_kind).end_range_tombstone();
+            ++_range_tombstones->begin;
+            if (need_preempt()) {
```
Instead of copy+pasting this serializer, can we instead just check add `if (thread::running_in_thread() && need_preempt())` to the existing serializer?

bhalevy replied:
`running_in_thread` is not absolutely reliable since it may return `true` in continuation-based async function or coroutines when they are called from a thread and before they yield (a.k.a schedule) the first time.
we can instead pass is_preemptible flag flag though, that would be based on the entry point.

bhalevy replied:
To eliminate any overhead of the extra logic maybe we can use a bool template parameter for IsPreemptible, but I'm not sure it's worth it.

bhalevy replied:
You know what, since we already have the `Output` template parameter we won't introduce a new template anyhow so I'll do it this way

avikivity commented on replica/database.hh:
```diff
@@ -441,8 +441,7 @@ private:
 
     uint64_t _failed_counter_applies_to_memtable = 0;
 
```
> This patch unfreezes all frozen mutations at the
(async) table::apply layer where we can yield if needed
and then eliminates the memtable apply path, given

Likely to regress performance, on top of the coroutinization.

avikivity commented on mutation/mutation_partition.cc:
```diff
@@ -478,6 +484,22 @@ mutation_partition::apply(const schema& s, mutation_partition_view p,
+    apply_resume res;
+    while (apply_monotonically(s, std::move(p2), no_cache_tracker, app_stats, is_preemptible::yes, res) == stop_iteration::no) {
+        co_await yield();
+    }
```
There's a good chance that this regresses writes.

We may want to fast-path small mutations.

Also, it can break atomicity for memtable/cache updates. While we yield, we see a partial result.

A solution can be to lock the entire cache/memtable, but it's hairy.

bhalevy replied:
It's not used on the hot path.
I added a comment in a later version noting that it's not suitable when apply must be atomic.

bhalevy commented:
In v3 ([fab16dd](https://github.com/scylladb/scylladb/commit/fab16ddff179009551b4b2f0ac0f0fd069f774e1)):
- rebased
- mutation_partition: apply_monotonically: do not support schema upgrade
  - cleaned up patch, FIXME comments
- mutation_partition: add apply_gently
  - added function comment documenting (lock of) atomicity aspects
- mutation_partition_serializer: add write_in_thread
  - used template `bool Preemptible` parameter to `write_serialized` and friends that has zero overhead when called from async path, and it is applied only when called from a thread.
- schema_tables: redact_columns_for_missing_features: upgrade mutation in place
- schema_tables: convert_schema_to_mutations: return mutations rather than canonical_mutations
- mutation: add clear_gently
- migration_manager: migration_request: run in seastar thread
  - used clear_gently
- tablets: read_tablet_mutations_in_thread
  - refactored out of previous patch for better reviewability
- storage_service: raft_pull_snapshot: run in seastar thread
- database: add unfreeze_and_apply_in_memory
  - used only for large (> 128KB) frozen mutations
  - verified that has negligible effect on perf-simple-query --write
  - verified that when used effect is manifested with more allocations but inst/op and tasks/op has only <1% overhead
- raft: group0_state_machine: write_mutations_to_database: use to_mutation_gently
- raft: group0_state_machine: merge_and_apply: spawn a seastar thread
- raft: group0_state_machine: write_mutations_to_database: run in seastar thread
- raft: group0_state_machine: write_mutations_to_database: freeze mutations in thread
  - deal with yet another stall-site related to large system mutations

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/canonical_mutation_test
:small_blue_diamond: boost/frozen_mutation_test
:small_blue_diamond: boost/mutation_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9272/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1251/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/)
#### Build Details:
- Duration: 2 hr 44 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/canonical_mutation_test
:small_blue_diamond: boost/frozen_mutation_test
:small_blue_diamond: boost/mutation_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9275/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1254/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/)
#### Failed Tests (1/30499):
- [test_simple_decommission_node_while_query_info[2]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8370/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_decommission_node_while_query_info_2_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_decommission_node_while_query_info[2])
#### Build Details:
- Duration: 2 hr 51 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/canonical_mutation_test 🔹 boost/frozen_mutation_test 🔹 boost/mutation_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9275/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1254/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8370/)
> 
> #### Failed Tests (1/30499):
> * [test_simple_decommission_node_while_query_info[2]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/8370/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_decommission_node_while_query_info_2_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_decommission_node_while_query_info%5B2%5D)
> 
> #### Build Details:
> * Duration: 2 hr 51 min
> * Builder: spider1.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/17903

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/canonical_mutation_test 🔹 boost/frozen_mutation_test 🔹 boost/mutation_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9272/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1251/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8361/)
> 
> #### Build Details:
> * Duration: 2 hr 44 min
> * Builder: spider7.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/18457

Apparently unrelated to this change and doesn't reproduce for me locally

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8382/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8382/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1260/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9281/)
#### Build Details:
- Duration: 1 hr 57 min
- Builder: spider2.cloudius-systems.com

denesb commented on replica/database.hh:
```diff
@@ -1514,6 +1514,7 @@ public:
     const gms::feature_service& features() const { return _feat; }
     future<> apply_in_memory(const frozen_mutation& m, schema_ptr m_schema, db::rp_handle&&, db::timeout_clock::time_point timeout);
     future<> apply_in_memory(const mutation& m, column_family& cf, db::rp_handle&&, db::timeout_clock::time_point timeout);
+    future<> unfreeze_and_apply_in_memory(const frozen_mutation& m, schema_ptr m_schema, column_family& cf, db::rp_handle h, db::timeout_clock::time_point timeout);
```
If there are no other callers planned for this method, let's just inline it into its single caller. The database API is crowded enough already.

bhalevy replied:
@denesb the problem is that the caller uses continuations and we don't to pessimize it and makexthe whole function a coroutine.
If only this slow path would be inlined we'll either need to use do_with or have a coroutine lambda function which I'm not sure is simpler than this method (which is a coroutine)

denesb replied:
I don't think we should go as far as extract arbitrary pieces of code into methods, just to make them coroutines. This is two lines of code, in-lining as continuations is not that bad.

bhalevy replied:
Ixm afraid that the continuations may add an allocation for do_with on top of the additional tasks/allocation if unfreeze_gently returns an unavailable future (as expected), therefore a coroutune should be slightly more efficient.

bhalevy commented on db/schema_tables.cc:
```diff
@@ -851,8 +850,8 @@ future<std::vector<canonical_mutation>> convert_schema_to_mutations(distributed<
             if (is_system_keyspace(partition_key)) {
                 continue;
             }
-            mut = redact_columns_for_missing_features(std::move(mut), features);
```
@denesb the problem is that the caller uses continuations and we don't to pessimize it and makexthe whole function a coroutine.
If only this slow path would be inlined we'll either need to use do_with or have a coroutine lambda function which I'm not sure is simpler than this method (which is a coroutine)

avikivity commented:
I don't see perf reports.

avikivity commented on mutation/mutation_partition.cc:
```diff
@@ -479,6 +481,21 @@ mutation_partition::apply(const schema& s, mutation_partition_view p,
 }
 
+future<>
+mutation_partition::apply_gently(const schema& s, mutation_partition_view p,
```
So far mutation_partition was a future-oblivious building block.

Perhaps we should keep it this way and make these apply_gently external utilities in separate files.

bhalevy replied:
okay, I'll try refactoring out to a separate file

denesb replied:
An alternative is to make a generic building block, which can drive the `apply_monotonically()` to completion, i.e. call any method which takes `is_preemptible` and invoke it in a loop, until it finishes, yielding in-between.

bhalevy replied:
@denesb we already have that: see apply_monotonically

denesb replied:
Yes. What I mean is that we can have wrapper which can drive such a method to completion, exposing a `future<>` based interface to the caller. This can live in a separate file, like Avi suggested.

bhalevy replied:
I'm adding mutation/async_utils in the next version

avikivity commented on mutation/canonical_mutation.cc:
```diff
@@ -67,6 +67,34 @@ mutation canonical_mutation::to_mutation(schema_ptr s) const {
 }
 
+future<mutation> canonical_mutation::to_mutation_gently(schema_ptr s) const {
+    auto in = ser::as_input_stream(_data);
```
Ditto, we may want to make this an external utility.

bhalevy replied:
okay, I'll try refactoring it out to a separate file

avikivity commented on mutation/mutation_partition_serializer.cc:
```diff
@@ -7,6 +7,8 @@
  */
 
+#include <seastar/core/preempt.hh>
+
```
> mutation_partition_serializer: add write_in_thread
> For stall-free freezing in a thread.

Why a thread?

bhalevy replied:
because the `ser::writer_of_mutation<bytes_ostream>` generated serialization code is synchronous.
See for example freeze_in_thread:
```
    auto fm = frozen_mutation(m.key());
    mutation_partition_serializer part_ser(*m.schema(), m.partition());

    ser::writer_of_mutation<bytes_ostream> wom(fm._bytes);
    std::move(wom).write_table_id(m.schema()->id())
                .write_schema_version(m.schema()->version())
                .write_key(m.key())
                .partition([&] (auto wr) {
                    part_ser.write_in_thread(std::move(wr));
                }).end_mutation();
```

Maybe we could generate async serialization boilerplate code in addition to the synchronous entry points
and then we would not need to run this in a thread.

If we had async serialization boiler

avikivity replied:
Let's generate async serialization code then, otherwise we throw threads everywhere

avikivity commented on mutation/mutation_partition_serializer.cc:
```diff
@@ -145,13 +147,16 @@ auto write_row_marker(Writer&& writer, const row_marker& marker)
         row_tombstones.add().write_start(rt.start).write_tomb(rt.tomb).write_start_kind(rt.start_kind)
             .write_end(rt.end).write_end_kind(rt.end_kind).end_range_tombstone();
+        if (Preemptible && need_preempt()) {
+            yield().get();
```
Could use maybe_yield() instead of need_preempt().

avikivity replied:
And use `if constexpr` to highlight that it's compile-time.

bhalevy replied:
> Could use maybe_yield() instead of need_preempt().

yes, good idea

> And use if constexpr to highlight that it's compile-time.

Sure, no problem.

avikivity commented on db/schema_tables.cc:
```diff
@@ -773,19 +773,18 @@ schema_ptr scylla_table_schema_history() {
 #endif
 
 static
```
>schema_tables: redact_columns_for_missing_features: upgrade mutation …
…in place

> When needed, the input mutation can be upgraded
> in place instead being copied and modified,
> to be returned as the function's result value.

1. missing motivation
2. seems a regression in readabilty - an in/out variable is less clear at the call site than a return value

bhalevy replied:
The motivation is reducing moves, though they are cheap as they only move a unique_ptr holding the mutation data.
This was particularly weird, in my eyes in:
```
adjust_schema_for_schema_features(std::vector<mutation> schema, schema_features features) {
    for (auto& m : schema) {
        m = redact_columns_for_missing_features(std::move(m), features);
    }
    return schema;
}
```

Which clearly modifies the mutations in place (as are all other call sites, but there it's less obvious).

avikivity replied:
Why weird? It's clear what the input and output are.

avikivity commented on db/schema_tables.cc:
```diff
@@ -837,13 +837,14 @@ future<table_schema_version> calculate_schema_digest(distributed<service::storag
     return calculate_schema_digest(proxy, features, std::not_fn(&is_system_keyspace));
 }
 
```
>schema_tables: convert_schema_to_mutations: return mutations rather t…
>…han canonical_mutations

>Let the caller make the canonical mutations.
>The next patch will do that gently in a thread.

missing motivation

bhalevy replied:
the above is supposed to be the motivation - i.e. to let the caller make the canonical mutation in a thread.
I'll rephrase it to clarify the intent.

avikivity commented on replica/database.cc:
```diff
@@ -1808,13 +1808,22 @@ future<> database::apply_in_memory(const frozen_mutation& m, schema_ptr m_schema
 
     data_listeners().on_write(m_schema, m);
 
```
> median 1580850.76 tps ( 70.3 allocs/op,  14.2 tasks/op,   54240 insns/op,        0 errors)

12 extra allocations for 500 insns/ops? it doesn't add up.

bhalevy replied:
I'm not sure exactly where they come from.
I suspect it's due to unfreezing outside the context of the memtable allocating_section.
Is it possible that those allocation weren't accounted for here:
```
memtable::apply(const frozen_mutation& m, const schema_ptr& m_schema, db::rp_handle&& h) {
    with_allocator(allocator(), [this, &m, &m_schema] {
        _table_shared_data.allocating_section(*this, [&, this] {
            auto& p = find_or_create_partition_slow(m.key());
            mutation_partition mp(*m_schema);
            partition_builder pb(*m_schema, mp);
            m.partition().accept(*m_schema, pb);
```

This code effectively unfreezes the frozen_mutation into `mp`, the difference is that it's done inside the allocating_section.

avikivity replied:
The report doesn't make sense. If the number of allocations increases by 12, I expect a large increase in the number of instructions.

bhalevy replied:
I suspect it might be an accounting problem that is just exposed here, but I can't prove this theory yet.

bhalevy replied:
@avikivity the extra allocations count comes from the fact the unfreezing is done outside of the memtable `allocating_section` - since we count only seastar mallocs and not lsa allocs.

With the following patch perf-simple-query --write reports 8 more allocations:
```diff
diff --git a/replica/memtable.cc b/replica/memtable.cc
index 77d1a95f9a..25199b37d1 100644
--- a/replica/memtable.cc
+++ b/replica/memtable.cc
@@ -796,13 +796,13 @@ memtable::apply(const mutation& m, db::rp_handle&& h) {
 
 void
 memtable::apply(const frozen_mutation& m, const schema_ptr& m_schema, db::rp_handle&& h) {
-    with_allocator(allocator(), [this, &m, &m_schema] {
+    mutation_partition mp(*m_schema);
+    partition_builder pb(*m_schema, mp);
+    m.partition().accept(*m_schema, pb);
+    _stats_collector.update(*m_schema, mp);
+    with_allocator(allocator(), [&] {
         _table_shared_data.allocating_section(*this, [&, this] {
             auto& p = find_or_create_partition_slow(m.key());
-            mutation_partition mp(*m_schema);
-            partition_builder pb(*m_schema, mp);
-            m.partition().accept(*m_schema, pb);
-            _stats_collector.update(*m_schema, mp);
             p.apply(region(), cleaner(), *_schema, std::move(mp), *m_schema, _table_stats.memtable_app_stats);
         });
     });
```

avikivity replied:
It makes sense that there are 8 extra allocations. But then I'd expect around 1600 extra instructions. Maybe my expectations are incorrect, but something smells here.

bhalevy replied:
1600 extra instructions from trading lsa allocations with seastar allocations?

I'll try to count lsa allocations in perf-simple-query in addition to the seastar allocations so we can see the whole picture.

avikivity commented on service/raft/group0_state_machine.cc:
```diff
@@ -139,6 +139,7 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
 }
 
 future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merger) {
+  return async([this, &merger] {
```
I don't like this thread proliferation.

Can't we use non-thread gentle functions?

bhalevy replied:
I'll see what it takes to generate async serialization code, with which we can have gentle freeze / make_canonical_mutation implementations that don't require a thread.

avikivity commented on mutation/frozen_mutation.cc:
```diff
@@ -87,6 +87,26 @@ frozen_mutation::frozen_mutation(const mutation& m)
     _bytes.reduce_chunk_count();
 }
 
+frozen_mutation::frozen_mutation(const partition_key& key)
```
I mean, the changelog should explain why a thread rather than the natural approach

bhalevy commented:
In v4 ([1325c1e](https://github.com/scylladb/scylladb/commit/1325c1edc164b1afeac34b17b74ffb3bf64c54f2)):
- test/perf: report also log_allocations/op
- mutation_partition: add apply_gently
  - moved to mutation/async_utils.{cc,hh}
  - as well as closely related async helpers introduced in this series
- idl-compiler: emit include directive in generated impl header file
- canonical_mutation: add to_mutation_gently
  - moved to mutation/async_utils
- idl-compiler: generate async serialization functions for stub members
  - to facilitate async freeze_gently and make_canonical_mutation_gently that do not require a thread
- mutation: add freeze_gently
  - in mutation/async_utils
  - does not require thread
- frozen_mutation: move unfreeze_gently to async_utils
- canonical_mutation: add make_canonical_mutation_gently
  - in mutation/async_utils
  - does not require thread
- schema_tables: redact_columns_for_missing_features: get input mutation using rvalue reference
  - kept signature "pipeline" design, just used rvalue reference for the input mutation to reflect what the function is for
  - and fixed `adjust_schema_for_schema_features` call site to move input mutation rather than copy it
- database: apply_in_memory: unfreeze_gently large mutations
  - updated perf-simple-query results that include also log_allocs/op
Dropped:
- addition proliferation of threads
- patch adding mutation::clear_gently (we may want it at some point but it's not strictly needed by this series)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8477/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8477/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/canonical_mutation_test
:small_blue_diamond: boost/frozen_mutation_test
:small_blue_diamond: boost/mutation_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8477/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1330/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9354/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8477/)
#### Build Details:
- Duration: 3 hr 9 min
- Builder: spider7.cloudius-systems.com

avikivity commented:
> addition proliferation of threads

Probably meant reduction

avikivity commented on test/perf/perf.cc:
```diff
@@ -12,12 +12,17 @@
 #include "seastarx.hh"
 #include "reader_concurrency_semaphore.hh"
 #include "schema/schema.hh"
+#include "utils/logalloc.hh"
```
> 79151.53 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53289 insns/op,        0 errors)

Looks like we're very wasteful. We should improve this sometime.

avikivity commented on mutation/mutation_partition.cc:
```diff
@@ -465,47 +465,34 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
-    apply_monotonically(s, std::move(p2), p_schema, app_stats);
+    if (s.version() != p_schema.version()) {
+        p2.upgrade(p_schema, s);
+    }
```
I wonder if we can drop this too. Since #2577 memtable/cache manage it themselves. The query pipeline can work with a single schema version (possible exception: when reading from memtable/cache).

avikivity replied:
Of course, not in this series.

avikivity commented on mutation/async_utils.hh:
```diff
@@ -0,0 +1,35 @@
+#include "mutation.hh"
+
+//
+// Applies p to current object.
```
s/current/target/

bhalevy replied:
Right.  copy-pasted from the respective `apply` method.  Will fix.

avikivity commented on mutation/async_utils.cc:
```diff
@@ -0,0 +1,62 @@
+    apply_resume res;
+    while (target.apply_monotonically(s, std::move(p2), no_cache_tracker, app_stats, is_preemptible::yes, res) == stop_iteration::no) {
+        co_await yield();
+    }
```
Shouldn't this have a special case when the schemas match?

bhalevy replied:
You mean for correctness, or for performance?
The implementation here just follows `void mutation_partition::apply(const schema& s, mutation_partition_view p, const schema& p_schema, mutation_application_stats& app_stats)`.

In theory, it might be possible to combine `p.accept_gently` using the `partition_builder` with `apply_monotonically` to apply the `mutation_partition_view` directly into `target` if the schemas match, so we won't need this temporary `partition_builder`, but it's not trivial to say the least.  We should look at it as a possible optimization as a follow up.

avikivity commented on idl-compiler.py:
```diff
@@ -1093,6 +1093,15 @@ def add_param_writer_object(name, base_state, typ, var_type="", var_index=None,
+                return futurize_invoke(f, writer_of_{typename}<Output>(_out)).then([this] {{
+                    {set_command}
+                    return after_{base_state}__{name}<Output>{return_command};
+                }});
```
Should we emit a coroutine here? I guess not, there's a good chance that the first continuation will complete without yielding.

bhalevy replied:
Right.  That was my thinking too.

avikivity commented on test/boost/frozen_mutation_test.cc:
```diff
@@ -56,18 +55,28 @@ static future<> _test_freeze_unfreeze(UnfreezeFunc&& unfreeze_func) {
         }).get();
-    });
+      }
+    );
```
The new tests miss interoperability between gentle an ungentle versions. We could have freeze/unfreeze work, and freeze_gently/unfreeze_gently work, but have different serialized formats that don't interoperate.

bhalevy replied:
will fix in next version

bhalevy replied:
fixed in v5

avikivity commented on test/boost/canonical_mutation_test.cc:
```diff
@@ -29,7 +29,7 @@ SEASTAR_TEST_CASE(test_conversion_back_and_forth) {
 SEASTAR_THREAD_TEST_CASE(test_conversion_back_and_forth_gently) {
     for_each_mutation([] (const mutation& m) {
-        canonical_mutation cm(m);
+        auto cm = make_canonical_mutation_gently(m).get();
```
Again need to check that the ungentle version matches the gentle version, without relying on to_mutation_gently() that is being tested here. You're testing two independent things in one test.

bhalevy replied:
will fix in next version

bhalevy replied:
fixed in v5

avikivity commented on replica/database.cc:
```diff
@@ -1808,6 +1809,14 @@ future<> database::apply_in_memory(const frozen_mutation& m, schema_ptr m_schema
+                return cf.apply(m, std::move(h), timeout);
+            });
+        });
+    }
```
will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.

bhalevy replied:
> Since there's a do_with here, how about extracting it to a coroutine method?

That's exactly what I did in the previous version, bit @denesb had reservations about that since the database class has a proliferation of methods already so he suggested to inline it.

> will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.

It might, since it's waiting for available memory, and the caller must keep the mutation alive.
```
    return dirty_memory_region_group().run_when_memory_available([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cg, holder = std::move(holder)]() mutable {
        do_apply(cg, std::move(h), m, m_schema);
    }, timeout);
```

avikivity replied:
> > Since there's a do_with here, how about extracting it to a coroutine method?
> 
> That's exactly what I did in the previous version, bit @denesb had reservations about that since the database class has a proliferation of methods already so he suggested to inline it.

Coroutinization is there to increase readability, and it can't be inlined since that makes the caller a coroutine too.


> 
> There was no difference in performance either way.

Of course, since you measure <128k writes.


> 
> > will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.
> 
> It might, since it's waiting for available memory, and the caller must keep the mutation alive.
> 

I meant, yield while applying a large mutation. It will probably stall.


> ```
>     return dirty_memory_region_group().run_when_memory_available([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cg, holder = std::move(holder)]() mutable {
>         do_apply(cg, std::move(h), m, m_schema);
>     }, timeout);
> ```

bhalevy replied:
> > > Since there's a do_with here, how about extracting it to a coroutine method?
> > 
> > 
> > That's exactly what I did in the previous version, bit @denesb had reservations about that since the database class has a proliferation of methods already so he suggested to inline it.
> 
> Coroutinization is there to increase readability, and it can't be inlined since that makes the caller a coroutine too.

There's a hybrid option as well, but I'm not sure it's worth it.
I'm ok wither way, so if you prefer a coroutine method, I'll add it back.
The hybrid option that saves the `do_with` is:
```
        return unfreeze_gently(m, std::move(m_schema)).then(coroutine::lambda([&cf, h = std::move(h), timeout] (auto m) mutable -> future<> {
            co_await cf.apply(m, std::move(h), timeout);
        }));
```

> 
> > There was no difference in performance either way.
> 
> Of course, since you measure <128k writes.

I commented out the other options and measured only this part - inline continuations vs. coroutine method.

> > > will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.
> > 
> > 
> > It might, since it's waiting for available memory, and the caller must keep the mutation alive.
> 
> I meant, yield while applying a large mutation. It will probably stall.

maybe I misunderstand you, but applying of the mutation is synchronous in table::do_apply.
The only place that table::apply may yield is `run_when_memory_available`

> > ```
> >     return dirty_memory_region_group().run_when_memory_available([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cg, holder = std::move(holder)]() mutable {
> >         do_apply(cg, std::move(h), m, m_schema);
> >     }, timeout);
> > ```

avikivity replied:
Mixed continuations/coroutines are a great way to obfuscate the code.

avikivity commented on service/raft/group0_state_machine.cc:
```diff
@@ -116,22 +117,25 @@ static bool should_flush_system_topology_after_applying(const mutation& mut, con
 
 static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_address from, std::vector<canonical_mutation> cms) {
-    std::vector<mutation> mutations;
+    std::vector<std::pair<schema_ptr, frozen_mutation>> mutations;
```
We have frozen_mutation_and_schema in db/view/view_update_generator.hh. Can extract it to frozen_mutation.hh.

bhalevy replied:
good idea, will do in next version.

bhalevy replied:
Hmm, in fact it was there all along since 9e14412528d

avikivity commented on service/raft/group0_state_machine.cc:
```diff
@@ -116,22 +117,25 @@ static bool should_flush_system_topology_after_applying(const mutation& mut, con
-    co_await proxy.mutate_locally(std::move(mutations), tracing::trace_state_ptr());
+    co_await coroutine::parallel_for_each(mutations, [&] (auto& x) {
+        return proxy.mutate_locally(x.first, x.second, tracing::trace_state_ptr(), db::commitlog::force_sync::no);
+    });
```
Maybe we should do this in storage_proxy(vector\<mutations>).

bhalevy replied:
can do, it makes sense

bhalevy replied:
done in v5

avikivity commented:
Looks good, some minor comments.

bhalevy commented:
> > addition proliferation of threads
> 
> Probably meant reduction

hmm, there was a "Dropped" headline above it the github formatting seem to have hid.
I edited the comment to make that clearer.

bhalevy commented:
In v5 ([890b890](https://github.com/scylladb/scylladb/commit/890b890e36a9e146b7c36d4a79d4e27182e4abfd)):
- rebased
- mutation_partition: add apply_gently
  - fixed code comments
- canonical_mutation: add to_mutation_gently
  - extended test case to be able to test all 4 combinations of sync/async to_mutation/make_canonical_mutation
- mutation: add freeze_gently
  - extended test case to be able to test all 4 combinations of sync/async freeze/unfreeze
- raft: group0_state_machine: write_mutations_to_database: freeze mutations gently
  - reused struct frozen_mutation_and_schema
- storage_proxy: add mutate_locally(vector<frozen_mutation_and_schema>) method
  - new patch

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8486/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8486/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/canonical_mutation_test
:small_blue_diamond: boost/frozen_mutation_test
:small_blue_diamond: boost/mutation_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8486/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9364/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1341/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8486/)
#### Build Details:
- Duration: 3 hr 25 min
- Builder: spider7.cloudius-systems.com

avikivity commented on service/raft/group0_state_machine.cc:
```diff
@@ -16,6 +16,7 @@
 #include "mutation/async_utils.hh"
 #include "seastar/core/abort_source.hh"
 #include "seastar/core/on_internal_error.hh"
+#include "seastar/coroutine/parallel_for_each.hh"
```
Need to use <> includes for seastar

bhalevy replied:
Right. copy-pasta from above. I'll fix the adjacent includes too

bhalevy replied:
https://github.com/scylladb/scylladb/pull/18513

yaronkaikov commented:
merged before we branched out to 6.0 - removing label `backport/6.0`
